datafusion/optimizer/src/optimizer.rs

```diff
@@ -274,22 +249,78 @@ impl Optimizer {
     pub fn with_rules(rules: Vec<Arc<dyn OptimizerRule + Send + Sync>>) -> Self {
         Self { rules }
     }
+}
+
+/// Recursively rewrites LogicalPlans
+struct Rewriter<'a> {
```

Comment:
`datafusion/optimizer/src/optimizer.rs` has all the important changes (to use the TreeNode API and stop copying)